### ISSUE TITLE
Filetransfer causes contacts to lose visibility of each other briefly

### ISSUE THREAD

**Issue by LittleVulpix** [irungentoo/toxcore#1220]
Observed on latest uTox ( win64-0.2.i.exe ) as well as on Miranda-NG with latest toxcore, so it's a toxcore issue - not client-specific. Cannot say if it was present previously...

Sending a LARGE (600-900MB) file to a contact can (sometimes) cause the contact to briefly appear offline. This causes an interruption of the filetransfer. uTox is able to resume the filetransfer automatically but I doubt this is a desired behavior. Not quite sure what is causing this because I was able to reproduce it with a friend who lives far away, as well as locally by sending a large file to utox running on a virtual machine; so it is not a routing problem as the traffic was going via vmware's LAN.

**Comment by aaannndddyyy** [comment]
I can confirm that. 
Not only large file transfers. Mainly on mobile connetiosn or very slow internet connections in some countries you get disconnected from your peer when doing a file transfer. This has happened for at least half a year, and it starst even at file sizes of a few hundred KB in some cases, so that even transferring a photo needs 3 resumes.

**Comment by LittleVulpix** [comment]
This is still present and rather upsetting considering that when you are transferring, you can't really talk to anyone as it keeps disconnecting you (and reconnecting) due to the transfer killing the line

**Comment by zetok** [comment]
Can confirm. Happens when connection isn't of best quality.

Testing with https://github.com/tylertreat/Comcast may be useful here? (and probably with other issues too?)
It would be really good if developers did run Tox clients with that on whole time, I think that this would allow them to experience most (if not all) of user problems with Tox network..

**Comment by LittleVulpix** [comment]
Actually here it happens even on a really, really good connection (fibre optics, 240/20Mbit). The transfer just goes too fast and then causes the protocol to go offline temporarily.

I spoke to irungentoo about this and showed him some code where another developer for the miranda plugin solved the issue (and I can confirm that). Irungentoo said he sees the problem and that he may be able to fix it in core, so we just need to wait :)

**Comment by aaannndddyyy** [comment]
This is still an issue with latest core.

**Comment by LittleVulpix** [comment]
@irungentoo @aaannndddyyy I can confirm this. Similarly to the "friends that appear offline even though they are online", this also generally happens to people with slower connections or connections with higher latency; although I've also once managed to disconnect a friend who has a 120/10 line. This was working without a hitch on the latest non-new-api toxcore previously and our dev tweaked it a little by increasing some... "chunk resend delay" or something similar.

It just feels like when transferring, tox waits for replies to acknowledge transferred data but the transfer itself disrupts the connection. It happens on my client (miranda NG) when transferring stuff to people (I have a very, very fast line). Is there something new/specific that needs to be told to toxcore so as to not choke the whole connection? This doesn't happen as much with uTox<->uTox transfer, for example, but I don't believe toxcore should even allow the client to be able to do something like that. Can you please help out with what would be the main culprit for such issues?

**Comment by homotopycolimit** [comment]
can confirm. file transfers are really fast, then stall and connection is lost.
connection is re-established. file transfer resumes. stalls again. repeat.

**Comment by LittleVulpix** [comment]
In worse scenarios, the file transfer never resumes (happened twice on utox v utox, on very slow connection of the uploader)

**Comment by LittleVulpix** [comment]
https://github.com/irungentoo/toxcore/commit/a0e783213bff0ad39dc1c1acb7e552f3f8b94917 - I'm currently testing my client with the new tox core along with a few other clients that have the updated core in it, and I have to say that it has improved this issue a lot (I find that strange as the change was so minor); but it's not placebo. I reverted back to the previous core and was immediately able to trigger a disconnect.

In one case I was able to trigger a disconnect with the above fix already in, but upon switching to tcp-only mode, the file was transferred successfully (previously file transfer was impossible regardless of the networking settings with this contact.)

I will test this further and if the issue doesn't repeat itself, I'll consider it resolved and close this issue.

**Comment by LittleVulpix** [comment]
Seems all the clients have been updated with this, and I can say that it's now much better. I'm closing this for now. I believe this was a regression after the migration to new api, but has since been remedied by https://github.com/irungentoo/toxcore/commit/a0e783213bff0ad39dc1c1acb7e552f3f8b94917

**Comment by aaannndddyyy** [comment]
it is still happening with latest core and both contacts on µtox

**Comment by LittleVulpix** [comment]
@aaannndddyyy it hasn't happened to me since the update actually, but irungentoo said that it could still happen, just that it's going to be a lot less likely. Should I reopen this?

**Comment by aaannndddyyy** [comment]
i happens for me with every file > 1 MB with a contact on a slow connection. since it reconnects, it resumes (on utox, with qrtox we had resume problems too), but stzill it is a nuissance when you are in a conversation. So yes, please reopen

**Comment by aaannndddyyy** [comment]
This ticket, #1349, lack of offline messages, extremely high ressource usage on phone ... these are my main usability issues with tox. Once they are resolved, I can happily use it with my collegues for daily conversation without nuissance.

**Comment by LittleVulpix** [comment]
@irungentoo @aaannndddyyy , reopened.

**Comment by aaannndddyyy** [comment]
@LittleVulpix thanks

**Comment by LittleVulpix** [comment]
It happened to me once today after a long time. In general I have these issues if I try to send something to someone after I've been online for an extended period of time.

**Comment by aaannndddyyy** [comment]
still happens with latest core. A 2 MB file caused 20 disconnects. It's really no fun.

But an obersation: utox is more affected by this than qtox, it seems.

The user in the first column has fast internet, the user in the seocnd column a rather slow one
utox : utox   -  many disconnects already after a few seconds
utox : qtox   -  many disconnects 
qtox : qtox   -  no disconnects at all even with 200+ MB if the fast peer sends. 
                    - almost immediate disconnects if the slow peer sends

The slow peer on qtox can also receive files from both, utox and qtox, simultaneously, which, as expected, results in the connection between utox and qtox disconnecting and the qtox qtox connetion staying intact and transferring files.

**Comment by aaannndddyyy** [comment]
workaround: For a 3MB file, the SENDER (not receiver) has to pause the transfer 7 times and give it some seconds to "take a breath". This way I was able to receive a 3MB file on qtox.
(this was tested only on qtox, not utox. As noted above utox suffers more from that, so it would probably need more pauses, maybe every two seconds)
toxcore is so busy with the filetransfer that not even the receipt confirmations of chat messages are getting through - they only arrive once the transfer was paused.

**Comment by aaannndddyyy** [comment]
still an issue as of October 20.

**Comment by GrayHatter** [comment]
@aaannndddyyy coming SOOONNNN™

**Comment by aaannndddyyy** [comment]
Hope that the info provided here helps diagnosing and finding the root cause.

**Comment by g4jc** [comment]
Confirming on latest qtox client. Bug happens to me nearly every time.
 Let me know if you need any help debugging.

**Comment by aaannndddyyy** [comment]
New api got merged, bug still present.

**Comment by LittleVulpix** [comment]
@aaannndddyyy new api was only regarding AV (audio/video, so just calls). No changes to filetransfers as of yet. :(

**Comment by aaannndddyyy** [comment]
OK. Didn't know that. I thought it was an overall revamp, so I thought posting an update on the status of this bug might be a good idea. Interesting fact is that in one directions it seems to work pretty fine, whereas in the other direction (same peer) it is hardly possible to transfer a 2MB file.

**Comment by GrayHatter** [comment]
https://gist.github.com/cleverca22/58710def3796638b1482

thanks @clever22

**Comment by GrayHatter** [comment]
and here https://camo.githubusercontent.com/48aecb1138b730951daa01d89b96299ba4acefbf/687474703a2f2f692e696d6775722e636f6d2f6e464c467969542e706e67

**Comment by irungentoo** [comment]
https://github.com/irungentoo/toxcore/commit/7022cef26edfccd2d4a5d47d3b7179958af7ce6a

Does this improve things?

**Comment by LittleVulpix** [comment]
@irungentoo I will do some testing tonight when Miranda NG tox maintainer updates the core.
@aaannndddyyy it would be awesome to get some test from you too, if you can grab the latest core.

**Comment by ovalseven8** [comment]
Just tested it with two instances of latest qTox (https://github.com/tux3/qTox/commit/5b42b9984b52fb5ec19a8c144086c19b9061e7b6) and unfortunately, it's not fixed.
One instance used normal UDP traffic and the other instance used Tor (TCP). File transfer failed. About half of the file (~ 33 MB) was transmitted.

**Comment by GrayHatter** [comment]
@ovalseven8, where'd you get your qTox and your toxcore/

**Comment by aaannndddyyy** [comment]
@irungentoo: It is not fixed. But it was behaving better this time. Not sure if it is arandom fluctation, but I gues it is not. So it seems you are moving in the right direction.

further info:
when I sent a big file it worked also before, when i sent many smaller files at once, the first once succeeded, maybe even second one, but then it used to break. Receiving was imposiblle even for small files unless they were manually paused every 200 KB or so.

Tried the same now. 
many files are working better now, and also receiving. 
But it is not fully fixed yet.
Also, there is a message sending lag during file transfers - but with bad network that seems reasonable. However I think, the text show have priority over the file. 
With calls it should be the other way round: written text need not arrive instantaneously but spoken word should. 

Many text messages are received repeatedly by the party who is receiving the file.

**Comment by irungentoo** [comment]
@aaannndddyyy 

https://github.com/irungentoo/toxcore/commit/0d0b74f5a8662a0b3b7c4da3537d1e8942f20656

Does this improve things further?

**Comment by aaannndddyyy** [comment]
@irungentoo it's hard to tell. 
It seems to further improve it, but still no fix.
I have no exact metric, I simply try. And if it fails after a few seconds, I know it gets back to how it was. This is not the case. It is transferring data fine for some time, then breaks. Both directions, from fast to slow contact and from slow to fast. My impression is that it lasted longer than with the previous fix.

**Comment by irungentoo** [comment]
https://github.com/irungentoo/toxcore/commit/6b7f1deb7929c4134fd70eb97bf2c1263fab0652

What about now?

**Comment by aaannndddyyy** [comment]
not fixed yet. It's hard to tell if it is better or same as previous. Network conditions vary. But what I know is, it is still disconnecting us. Disconnects do not happen if no file transfer is underway, so it's not that the entire internet goes down for one of us.

**Comment by irungentoo** [comment]
@aaannndddyyy What is the speed and latency of your internet and the internet of the people you try to send files to?

**Comment by g4jc** [comment]
I am still having issues as well. I would also mention that the person I attempted to share files too is on broadband with a great line, while I am using a jittery wireless connection.

One thing both of us noticed was there is a huge speed increase by about 20/30kbps just prior to me being disconnect whether it be upload or download. Not sure if this is useful or not, but figured I'd report it nonetheless. It almost seems like a packet race condition if such a thing is possible.

As another person on the thread mentioned - You can also use comcast, trickle, and similar tools to help simulate this problem.

Thanks for your persistence on trying to squash this bug! :1:

**Comment by aaannndddyyy** [comment]
1 - 2.5Mbit/s up,
2-4.0 down,
60-205ms ping,
140-700 Jitter

and

   0.4  up,
   0.3  down,
   30 -105 ping,
   450-520 Jitter

**Comment by irungentoo** [comment]
What about now on latest core?

**Comment by aaannndddyyy** [comment]
Just grabbed binary from tox.chat - still not fixed. :(

**Comment by GrayHatter** [comment]
@aaannndddyyy that binary is still out of date...

best way to be sure would be to recompile toxcore, and then your client your self

**Comment by aaannndddyyy** [comment]
@GrayHatter Thanks for the info. I am not in a place where could compile, right now, so I guess I'll wait until they update. This is done automatically, I assume. Is there a way to know when they are up to date?

**Comment by TheNain38** [comment]
I've seen that at each new toxcore commit the client is recompiled automatically.

**Comment by GrayHatter** [comment]
@aaannndddyyy, no I lied... you're likely using the correct build.

**Comment by towlie** [comment]
I hope that you don't use the utox updater to test the latest commit if you use utox.

**Comment by aaannndddyyy** [comment]
https://build.tox.chat/view/Clients/job/qTox-qt5.4.2_build_linux_x86-64_release/ says built 50 mins ago, and last check in was an hour ago, so it seems it contains it

**Comment by aaannndddyyy** [comment]
it's actually even worse than yesterday. but external factors could play a role. contact reports heavy weather

**Comment by irungentoo** [comment]
@aaannndddyyy if you try with uTox, does it still fail?

**Comment by aaannndddyyy** [comment]
@irungentoo: Yay, it seems to work better with utox!!  So congrats! :)
but no totally fixed. 
And in case it were fixed for utox and not for qtox, then:
Clients should actually just be interfaces with different look and user experience, but not different on the networking layer.
So if a client can mess up the connection, something is wrong. Except if a client directly says it wants to cancel the transfer or reset the connection, this should happen. In all other cases clients should simply not be **able** to mess up

**Comment by aaannndddyyy** [comment]
there also sems to be a bug with resuming. the file was 25% done when we got disconnected and this time it started again from zero, whereas the other before, resumed just fine.

**Comment by aaannndddyyy** [comment]
7 disconnects for 30MB of transferferred file utox - utox

**Comment by GrayHatter** [comment]
@aaannndddyyy can you provide any information on the file not resuming with uTox?

**Comment by aaannndddyyy** [comment]
Maybe it is related to this bug and not relly a resume bug.

I send a file.
Progress bar advances and shows ca 25% done.
Then we lose connection. 
Then file restarted at 0. But the other one did resume well. So maybe it has sent iut 2.5MB that were never received, and showed them in the progress , whereas after the disconnect the other side told it had received zero?

Anyway, the huge paper clip is kinda ugly ;)

GrayHatter <notifications@github.com> hat geschrieben:

@aaannndddyyy can you provide any information on the file not resuming with uTox?

**Comment by LittleVulpix** [comment]
@irungentoo the latest toxcore, I managed to send a file to my friend at 3.12MB/s (the max speed I can send a file at). Then he tried to send me one, and it was going at 100-300kB/s for a bit until it DCd us. (we both have 300/30Mbit line, same country). The transfer from him was tcp relayed because it was coming from a different IP than his. Either way, not fixed..

Side question - whose "fault" it is when transfer is udp in one direction but tcp in the other? i.e. whose ports are not reachable via udp? We both have public IP and we both have tcp+udp ports 33445-33450 forwarded to our computers...

**Comment by irungentoo** [comment]
Can you reproduce this on latest core? Note that this time both the sender and receiver need to have latest core.

**Comment by towlie** [comment]
I suggest you not to test this through the packages repository. I have the feeling that the binaries from there are not always up to date.
Better test with one these binaries: 
Ubuntu users: 
https://build.tox.chat/view/Clients/job/qTox-unity_build_linux_x86-64_release/
https://build.tox.chat/view/Clients/job/qTox-unity_build_linux_x86_release/
or
https://build.tox.chat/view/Clients/job/qTox-qt5.4.2_build_linux_x86-64_release/
https://build.tox.chat/view/Clients/job/qTox-qt5.4.2_build_linux_x86_release/
or
https://build.tox.chat/view/Clients/job/qTox_build_linux_x86-64_release/
https://build.tox.chat/view/Clients/job/qTox_build_linux_x86_release/

**Comment by LittleVulpix** [comment]
@irungentoo Tested with latest toxcore (both clients). Transferred about 80MB of a 1GB file, then the speed went to 0, he DCd briefly - again the same thing. TCP mode, was going through a relay. I.e. no change.

**Comment by GrayHatter** [comment]
@irungentoo would the relay also need to be on the newest core?

**Comment by irungentoo** [comment]
@GrayHatter  no, only the two sides.

**Comment by LittleVulpix** [comment]
Either way, it feels like the algorithm tries to go too fast too quickly.

Using UDP, my friend has 400kB/s download speed. The speeds (in kB/s) go like this:
1s - 20
2s - 35
3s - 60
4s - 130
5s - 240
6s - 450
7s - 640 (already obviously not possible)
8s - 930 (nor is this)
9s - 930 (progress bar stuck showing mo more data is actually being sent)
10s - 0
11s - 0
12s - 0
13s - 0
14s - broken transfer

**Comment by irungentoo** [comment]
https://github.com/irungentoo/toxcore/commit/994a5db1058119409005a3874c936ed6c0b49766

Does this improve things?

**Comment by aaannndddyyy** [comment]
Yes.
Looking much better. It's definitely useable now. At least in the one limited test I could give it. It disconnected twice, so not entirely fixed, but much better. Thanks a lot.

**Comment by ovalseven8** [comment]
> It disconnected twice

Nonetheless, this shouldn't happen.

**Comment by LittleVulpix** [comment]
@irungentoo before I begin, did you do something to how reconnecting to the network works? Both me and my friend reconnected within like 10 seconds as opposed to the usual 30-60 seconds when we were switching from UDP to TCP....

Anyway, to the point; this is miles better. Tested it with one friend so far, we would always get a disconnect and wonky speed, but in UDP mode, it went as fast as possible in both directions (I was sending at 3MB/s, my friend at about 1.5MB/s), and most importantly, the transfer was stable; the speed wasn't going from 0 to 3MB/s, but it was just hovering around 2.9-3.1MB/s. It was perfect.

For TCP mode, it worked from me to him. Speed was a little wobbly (around 900kB/s to 1.5MB/s) but that is because of the relays of course. When we tried from him to me, it was going okay for a while, the speed was actually much more stable (constantly around 1.3-1.5MB/s - my friend's max upload speed), but then it went down to 300, then to 100 and 0 and then he disconnected again. I assume it's again the congestion, causing him to be unable to respond to a ping and then getting booted off the network.

Definitely a major improvement. UDP was perfect, TCP was... better, hehe. It's very close to being fixed, I'm sure. I will test it with one more friend whose connection is way worse and he lives much further away, but so far I'm really happy.

**Comment by LittleVulpix** [comment]
Alright! Now tested with my friend from far away (his line is 4Mbit/0.5Mbit). UDP mode From me to him=> speed around 290-350kB/s, perfect. From him to me, speed around 29-40kB/s, perfect too. No failures and we could also talk while transferring which previously didn't work very well! Major improvement.

We did not test TCP mode because it failed with the previous friend already so obviously there is a problem there, but UDP seems to be flawless (for me anyway!)

**Comment by reselver** [comment]
Unfortunately, it's not fixed yet. Discovered some bugs with latest toxcore:
@irungentoo 

- [ ] 1. When I open two Tox instances and make a file transfer between the two instances, I only get a speed maximum around 18 MiB/s. That's pretty slow for a direct connection. Much more should be possible, shouldn't it?

- [ ] 2. If I make a file transfer with one or more contacts during an audio or video call, the call quality gets worse and sometimes the audio/video hangs for short moments. Normally, the call should still working great and the file transfers are slowed down in return.

- [ ] 3. Did a file transfer during a call and the file transfer failed and even the call ended, eventually. The output was: *Warning: Call with friend 0 died of unnatural causes!*
Probably related to the problem above.

**Comment by GrayHatter** [comment]
@reselver what client are you using, and where'd you get your version of client and core?

**Comment by g4jc** [comment]
Tested here with two clients running qtox. The problem appears to be resolved. Did not test during audio or video chats, but was able to transfer 15MB files without error.

Thanks!

**Comment by reselver** [comment]
@GrayHatter: Latest qtox - with latest toxcore. Can confirm that some bugs are solved now - was able to do file transfer without some disconnections. 

But file transfers and audio or video in combination is not great yet (as written). Call and file transfer failed or call is suffering. The three issues I wrote above are the last three bugs (for now) I could discover. Hope you can fix it because it's important.

**Comment by quininer** [comment]
@reselver  The latest is a bad description, you should give a specific version.

**Comment by reselver** [comment]
@quininer: https://github.com/irungentoo/toxcore/commit/73b2144edcfd1ca617e9054479b66ab0c0361a14

**Comment by irungentoo** [comment]
Can someone try a local transfer on windows and tell me what is the maximum speed that they get is.

**Comment by TheNain38** [comment]
@irungentoo It was around 30-40MiB/s for me

**Comment by LittleVulpix** [comment]
@irungentoo It was about 24-28MB/s for me between two toxes (gbit lan between two computers. In comparison, samba speed was at 100-110MB/s until disk i/o became an issue, then it went to about 80MB/s)

**Comment by towlie** [comment]
Was it ever faster? You should remember that your computer has to encrypt the content which requires more CPU time compared to unencrypted content.

**Comment by LittleVulpix** [comment]
@towlie It was faster on old toxcore, way back in the day. It didn't work at all on the recent toxcore so until now I didn't get to test it again. And I didn't notice some crazy CPU utilization on either side, these were somewhat powerful machines.

**Comment by LittleVulpix** [comment]
@GrayHatter @irungentoo @reselver @aaannndddyyy  Since the name and point of this issue was to make the file transfers reliable (and after a week of testing I have not yet had a failed file transfer with even the friends with the worst internet connection possible), I'm closing this issue as resolved.

@reselver If you are  bothered by the slow speed for LAN transfers, please make a new ticket
@irungentoo Thanks! I consider the file transfers fixed as per my "week of your time" request. Much appreciated!

Everyone else, thanks for valuable input during this long-overdue issue!

**Comment by aaannndddyyy** [comment]
@LittlerVulpix: Maybe you can pick your contact with the most unreliable 
connection (the one you were having most problems with), and then try a 
transfer to him and one from him (maybe in parallel) via tor (maybe even 
both on tor) in order to maximally stress test it. If it does not fail 
anymore, it really is fully resolved (And yes, I am aware that a tor 
connection could also break without toxcore being to blame, but if 
without transfers the connection is stable and with it is not, then it 
would be an indicator).

So far for me it is working fine, though I didn't have very much chance 
to test it.
Would be interested however, in case you perfomr that test, what the 
results were.

**Comment by LittleVulpix** [comment]
@aaannndddyyy I'm glad my friends didn't just quit on me during all the tests I put them through :D I tested it over tor with someone with a decent connection and I had no issues. I can't say how it will work with someone with really bad connection over tor... I think you'll have to test that :p but it's like, such an unlikely usage scenario... tox filetransfers now work even over super crappy lines where before they'd all fail even over good lines.

**Comment by GrayHatter** [comment]
if you're friend has a really bad connection, and is using Tox over Tor... it's going to break... that's just how it goes. Thankfully toxcore supports file resuming, so at least it has that going for it which is nice.

**Comment by aaannndddyyy** [comment]
@Grayhatter Toxcore shpuld not deliberately break a connection. So as lpng as the underlying transport is not interrupted, no matter how slow it is, toxcore should just work.

**Comment by g4jc** [comment]
I was able to reproduce the disconnect bug after a stress test on a slow internet connection. Starting 3 file transfers at the same time causes the same effect as the original bug report. On the bright side anyone using a slow internet connection probably won't be transferring multiple files at the same time.

**Comment by GrayHatter** [comment]
@g4jc where'd you get your version of toxcore, what client, where'd you get your client, what version of client? And all the same questions for your friend/peer.

**Comment by g4jc** [comment]
@GrayHatter qTox. We were running https://github.com/tux3/qTox/commit/db52ef6e6a843faa8543b4726a5c281b8f1e1d97

I may update again and see since it looks like another version is out already.

**Comment by alexeysvrv** [comment]
@tux3 
> I was able to reproduce the disconnect bug after a stress test on a slow internet connection. Starting 3 file transfers at the same time causes the same effect as the original bug report. On the bright side anyone using a slow internet connection probably won't be transferring multiple files at the same time. 
qTox. We were running tux3/qTox@db52ef6
I may update again and see since it looks like another version is out already.
